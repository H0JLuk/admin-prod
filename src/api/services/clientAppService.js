import { Api } from '../apiClient';
import { getReqOptions } from './index';

export async function getClientAppList() {
    return Api.get('/admin/clientApplication', getReqOptions());
}

export async function addClientApp(clientAppDto) {
    return Api.post('/admin/clientApplication', clientAppDto, getReqOptions());
}

export async function copyClientApp(clientAppDto) {
    return Api.post('/admin/clientApplication/copy', clientAppDto, getReqOptions());
}

export async function updateClientApp(id, clientAppDto) {
    return Api.put(`/admin/clientApplication/${id}`, clientAppDto, getReqOptions());
}

export async function deleteClientApp(id) {
    return Api.delete(`/admin/clientApplication/${id}`, getReqOptions());
}

export async function updateProperties(id, properties) {
    return Api.put(`/admin/clientApplication/${id}/properties`, properties, getReqOptions());
}

export async function reorderClientApp(idMap) {
    return Api.put('/admin/clientApplication/reorder', idMap, getReqOptions());
}

/** TODO: remove DEFAULT_APP and functions after second phase */
const DEFAULT_APP = 'DEFAULT_APP';

export function setDefaultAppCode(appCode) {
    sessionStorage.setItem(DEFAULT_APP, appCode);
}

export function getDefaultAppCode() {
    return sessionStorage.getItem(DEFAULT_APP) || '';
}

export function removeDefaultAppCode() {
    sessionStorage.removeItem(DEFAULT_APP);
}
