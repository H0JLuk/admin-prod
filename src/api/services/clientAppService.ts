import {
    DefaultApiResponse,
    ClientAppDto,
    DefaultCreateDtoResponse,
    ListResponse,
    SaveClientApp,
} from '@types';
import { Api } from '../apiClient';
import { getReqOptions } from './index';

export function getClientAppList() {
    return Api.get<ListResponse<ClientAppDto>>('/admin/clientApplication', getReqOptions());
}

export async function getActiveClientApps() {
    const { list = [] } = await getClientAppList();
    return list.filter(({ isDeleted }) => !isDeleted);
}

export function getClientAppInfo(appCode: string) {
    return Api.get<ClientAppDto>(`/admin/clientApplication/${appCode}`, getReqOptions());
}

export function addClientApp(clientAppDto: SaveClientApp) {
    return Api.post<DefaultCreateDtoResponse>('/admin/clientApplication', clientAppDto, getReqOptions());
}

export function copyClientApp(clientAppDto: SaveClientApp) {
    return Api.post<DefaultCreateDtoResponse>(
        '/admin/clientApplication/copy',
        clientAppDto,
        getReqOptions()
    );
}

export function updateClientApp(id: number, clientAppDto: SaveClientApp) {
    return Api.put<DefaultApiResponse>(`/admin/clientApplication/${id}`, clientAppDto, getReqOptions());
}

export function deleteClientApp(id: number) {
    return Api.delete<DefaultApiResponse>(`/admin/clientApplication/${id}`, getReqOptions());
}

export function reorderClientApp(idMap: { idMap: Record<number, number>; }) {
    return Api.put<DefaultApiResponse>('/admin/clientApplication/reorder', idMap, getReqOptions());
}

/** TODO: remove DEFAULT_APP and functions after second phase */
const DEFAULT_APP = 'DEFAULT_APP';

export function setDefaultAppCode(appCode: string) {
    sessionStorage.setItem(DEFAULT_APP, appCode);
}

export function getDefaultAppCode() {
    return sessionStorage.getItem(DEFAULT_APP) || '';
}

export function removeDefaultAppCode() {
    sessionStorage.removeItem(DEFAULT_APP);
}
