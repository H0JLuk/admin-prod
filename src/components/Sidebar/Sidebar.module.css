.menu {
    position: sticky;
    top: 0;
    width: 300px;
    max-height: 100vh;
    flex-shrink: 0;
    background-color: white;
    border-right: 1px solid #d4d9de;
}

.blockLogo {
    border-bottom: 1px solid #d4d9de;
    margin-bottom: 15px;
}

.logo {
    width: 45%;
    padding: 17px;
}

.menu__item {
    display: flex;
    text-decoration: none;
    font-family: SBSansText, sans-serif;
    font-size: 14px;
    font-weight: 600;
    line-height: 1.5;
    letter-spacing: -0.3px;
    background-color: transparent;
    padding: 10px 20px;
    cursor: pointer;
}

.menu__item:hover {
    background-color: #f8f8f8;
}

.menu__item > span {
    margin-right: 10px;
}

body:not(:global(.not_highlight_app)) .activeApp,
.active {
    color: #09a552 !important;
    background-color: #f8f8f8;
}

.app {
    color: #868d98;
    font-size: 11px;
    font-weight: 100;
    text-transform: uppercase;
    margin-top: 15px;
    pointer-events: none;
}

.addButton {
    color: #868d98 !important;
}

.menuTitle > div {
    background-color: rgba(0, 0, 0, 0.1);
    color: white;
}

.menu :global(.ant-menu.ant-menu-dark) {
    border-radius: 5px;
    background-color: transparent;
    border: 0;
    color: white;
}

.menu :global(.ant-menu-item) {
    height: unset !important;
    min-height: 40px;
    white-space: unset;
    margin: 0 !important;
    padding-left: 0px !important;
    padding-right: 0px;
}

.menu :global(.ant-menu-item::after) {
    border: none;
}

.menu :global(.ant-menu-item-selected) {
    color: #09A552 !important;
    background-color: #F8F8F8 !important;
}

.menu :global(.ant-menu-item:hover) {
    color: black;
}

.menu :global(.ant-menu-item a:hover) {
    color: black;
}
