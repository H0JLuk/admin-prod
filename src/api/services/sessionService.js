import { removeDefaultAppCode } from './clientAppService';

const TOKEN = 'dzo_token';
const ROLE = 'role';
const APP_CODE = 'appCode';

export function storeUserData(userToken, userRole) {
    window.sessionStorage.setItem(TOKEN, userToken);
    window.sessionStorage.setItem(ROLE, userRole);
}

export function deleteUserData() {
    window.sessionStorage.removeItem(TOKEN);
    window.sessionStorage.removeItem(ROLE);
    deleteAppCode();
    removeDefaultAppCode(); /* TODO: remove this after second phase */
}

export function getSession() {
    return window.sessionStorage.getItem(TOKEN);
}

export function getRole() {
    return window.sessionStorage.getItem(ROLE);
}

export function saveAppCode(appCode) {
    window.sessionStorage.setItem(APP_CODE, appCode);
}

export function deleteAppCode() {
    window.sessionStorage.removeItem(APP_CODE);
}

export function getAppCode() {
    return window.sessionStorage.getItem(APP_CODE);
}
