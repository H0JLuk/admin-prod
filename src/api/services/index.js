import { getSession } from './sessionService';

export const getReqOptions = () => {
    const token = getSession();
    return {
        headers: { token }
    };
};
