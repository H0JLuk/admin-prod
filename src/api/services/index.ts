import { ApiRequestHeaders, ApiRequestOptions } from '@types';
import { getDefaultAppCode } from './clientAppService';
import { getAppCode, getSession } from './sessionService';

const TOKEN_HEADER = 'token';
const APP_CODE_HEADER = 'clientAppCode';
const CONTENT_TYPE = 'Content-Type';
const ACCEPT = 'Accept';

export const getReqOptions = (contentType?: string, currAppCode?: string): ApiRequestOptions => {
    const token = getSession();
    const appCode = currAppCode || getAppCode() || getDefaultAppCode(); // TODO: remove getDefaultAppCode after second phase
    const headers = {} as ApiRequestHeaders;
    token && (headers[TOKEN_HEADER] = token);
    appCode && (headers[APP_CODE_HEADER] = appCode);
    contentType && (headers[ACCEPT] = '*/*');
    contentType && (headers[CONTENT_TYPE] = contentType);
    return { headers };
};
