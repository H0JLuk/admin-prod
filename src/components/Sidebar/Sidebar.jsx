import React, { memo, useEffect, /* useMemo, */ useState } from 'react';
import cn from 'classnames';
import { matchPath, NavLink, useHistory, useLocation } from 'react-router-dom';
import { Menu } from 'antd';
import { PROMO_CAMPAIGN_PAGES, ROUTE, ROUTE_ADMIN, ROUTE_OWNER } from '../../constants/route';
import { getClientAppList, setDefaultAppCode } from '../../api/services/clientAppService';
import { getAppCode, getRole, saveAppCode } from '../../api/services/sessionService';
import { goApp } from '../../utils/appNavigation';
import { resolveMenuItemsByRoleAndAppCode } from '../../constants/menuByRole';
import ROLES from '../../constants/roles';

import sberLogo from '../../static/images/sber-logo.png';

import styles from './Sidebar.module.css';

const routesForNonRender = [
    `${ ROUTE_ADMIN.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.ADD_PROMO_CAMPAIGN }`,
    `${ ROUTE_ADMIN.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_INFO }`,
    `${ ROUTE_ADMIN.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_EDIT }`,
    `${ ROUTE_ADMIN.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_COPY }`,
    `${ ROUTE_OWNER.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.ADD_PROMO_CAMPAIGN }`,
    `${ ROUTE_OWNER.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_INFO }`,
    `${ ROUTE_OWNER.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_EDIT }`,
    `${ ROUTE_OWNER.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_COPY }`,
];

const routesForApps = [
    `${ROUTE_ADMIN.APPS}`,
    `${ROUTE_OWNER.APPS}`
];

const Sidebar = () => {
    const history = useHistory();
    const { pathname } = useLocation();
    const [appsList, setAppsList] = useState([]);

    const appCode = getAppCode() || '';
    const role = getRole();
    const [menuItems] = resolveMenuItemsByRoleAndAppCode(role, appCode);

    useEffect(() => {
        if (role === ROLES.USER_MANAGER) {
            return;
        }

        (async () => {
            const { clientApplicationDtoList = [] } = await getClientAppList();
            const sortedDtoList = clientApplicationDtoList.filter(({ isDeleted }) => !isDeleted);
            /* Временный костыль чтобы можно было перемещаться по приложению без крашей */
            if (!getAppCode()) {
                setDefaultAppCode(sortedDtoList[0].code);
            }

            setAppsList(sortedDtoList);
        })();
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const handleAdministrate = (dzoItem) => {
        saveAppCode(dzoItem.code);
        goApp(history, role);
    };

    if (matchPath(pathname, { path: routesForNonRender })) {
        return null;
    }

    return (
        <div className={ styles.menu }>
            <div className={ styles.blockLogo }>
                <img className={ styles.logo } src={ sberLogo } alt="sberLogo" />
            </div>
            <Menu mode="inline">
                {menuItems.map((item) => (
                    <Menu.Item key={ item.label }>
                        <NavLink
                            key={ item.label }
                            to={ item.path }
                            className={ styles.menu__item }
                            activeClassName={ styles.active }
                        >
                            { item.label }
                        </NavLink>
                    </Menu.Item>
                ))}
                {!(matchPath(pathname, { path: routesForApps })) && (appsList || []).map((dzoItem, index) => (
                    <Menu.Item key={ dzoItem.id } className={ cn(styles.appLink, { [styles.firstApp]: !index }) }>
                        {/* TODO: change this to NavLink or add active className after we will know url for apps */}
                        <div
                            onClick={ () => handleAdministrate(dzoItem) }
                            className={ cn(styles.menu__item, { [styles.activeApp]: appCode === dzoItem.code }) }
                        >
                            { dzoItem.displayName }
                        </div>
                    </Menu.Item>
                ))}
                { role !== ROLES.USER_MANAGER && (
                    <Menu.Item key="oldDesign">
                        <NavLink to={ `${ ROUTE.OLD_DESIGN }${ pathname }` } className={ styles.oldDesignLink }>
                            Старый дизайн
                        </NavLink>
                    </Menu.Item>
                ) }
            </Menu>
        </div>
    );
};

export default memo(Sidebar);