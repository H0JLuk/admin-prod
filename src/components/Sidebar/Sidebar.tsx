import React, { memo, useEffect, useState } from 'react';
import { matchPath, NavLink, useHistory, useLocation } from 'react-router-dom';
import cn from 'classnames';
import { PROMO_CAMPAIGN_PAGES, ROUTE_ADMIN, ROUTE_OWNER } from '@constants/route';
import { getActiveClientApps, setDefaultAppCode } from '@apiServices/clientAppService';
import { getAppCode, getRole, saveAppCode } from '@apiServices/sessionService';
import { goApp } from '@utils/appNavigation';
import { resolveMenuItemsByRoleAndAppCode } from '@constants/menuByRole';
import ROLES from '@constants/roles';
import { ClientAppDto } from '@types';

import sberLogo from '@imgs/sber-logo.png';

import styles from './Sidebar.module.css';

const routesForNonRender = [
    `${ ROUTE_ADMIN.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.ADD_PROMO_CAMPAIGN }`,
    `${ ROUTE_ADMIN.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_INFO }`,
    `${ ROUTE_ADMIN.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_EDIT }`,
    `${ ROUTE_ADMIN.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_COPY }`,
    `${ ROUTE_OWNER.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.ADD_PROMO_CAMPAIGN }`,
    `${ ROUTE_OWNER.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_INFO }`,
    `${ ROUTE_OWNER.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_EDIT }`,
    `${ ROUTE_OWNER.PROMO_CAMPAIGN }${ PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_COPY }`,
];

const routesForApps = [
    `${ROUTE_ADMIN.APPS}`,
    `${ROUTE_OWNER.APPS}`,
];

const rolesWithoutClientApps = [
    ROLES.USER_MANAGER,
    ROLES.PARTNER,
];

const Sidebar = () => {
    const history = useHistory();
    const { pathname } = useLocation();
    const [appsList, setAppsList] = useState<ClientAppDto[]>([]);

    const appCode = getAppCode() || '';
    const role = getRole();
    const [menuItems] = resolveMenuItemsByRoleAndAppCode(role, appCode);

    useEffect(() => {
        if (rolesWithoutClientApps.includes(role)) {
            return;
        }

        (async () => {
            const clientAppList = await getActiveClientApps();
            /* Временный костыль чтобы можно было перемещаться по приложению без крашей */
            if (!getAppCode()) {
                setDefaultAppCode(clientAppList[0].code);
            }

            setAppsList(clientAppList);
        })();
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const handleAdministrate = (dzoItem: ClientAppDto) => {
        saveAppCode(dzoItem.code);
        goApp(history, role);
    };

    if (matchPath(pathname, { path: routesForNonRender })) {
        return null;
    }

    return (
        <div className={styles.menu}>
            <div className={styles.blockLogo}>
                <img className={styles.logo} src={sberLogo} alt="sberLogo" />
            </div>
            <ul className={styles.menuLinks}>
                {menuItems.map((item) => (
                    <li key={item.label}>
                        <NavLink
                            key={item.label}
                            to={item.path}
                            className={styles.menu__item}
                            activeClassName={styles.active}
                        >
                            {item.label}
                        </NavLink>
                    </li>
                ))}
                {!(matchPath(pathname, { path: routesForApps })) && (appsList || []).map((dzoItem, index) => (
                    <li key={dzoItem.id} className={cn(styles.appLink, { [styles.firstApp]: !index })}>
                        {/* TODO: change this to NavLink or add active className after we will know url for apps */}
                        <div
                            onClick={() => handleAdministrate(dzoItem)}
                            className={cn(styles.menu__item, { [styles.activeApp]: appCode === dzoItem.code })}
                        >
                            {dzoItem.displayName}
                        </div>
                    </li>
                ))}
            </ul>
        </div>
    );
};

export default memo(Sidebar);
