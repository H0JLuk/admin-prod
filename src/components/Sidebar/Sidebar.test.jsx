import React from 'react';
import { shallow } from 'enzyme';
import { fireEvent, render } from '@testing-library/react';
import { matchPath } from 'react-router-dom';
import { getActiveClientApps } from '../../api/services/clientAppService';
import { getAppCode, getRole, saveAppCode } from '../../api/services/sessionService';
import Sidebar from './Sidebar';
import { sleep } from '../../setupTests';
import { goApp } from '../../utils/appNavigation';
import { clientAppListTestResponse, resolveMenuItemsByRoleAndAppCodeValue } from '../../../__tests__/constants';
import * as menuRoleFuncs from '../../constants/menuByRole';

const mockPush = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        push: mockPush,
    }),
    useLocation: () => ({ pathname: '/admin/users' }),
    matchPath: jest.fn(),
    NavLink() {
        return <a>NavLink</a>;
    },
}));

jest.mock('../../api/services/sessionService', () => ({
    getAppCode: jest.fn(),
    getRole: jest.fn(),
    saveAppCode: jest.fn(),
}));

jest.mock('../../api/services/clientAppService', () => ({
    getDefaultAppCode: jest.fn(),
    getActiveClientApps: jest.fn(),
    setDefaultAppCode: jest.fn(),
}));

jest.mock('../../utils/appNavigation', () => ({
    goApp: jest.fn(),
}));

describe('<Sidebar /> tests', () => {
    it('Sidebar snapshot', () => {
        const SidebarItem = shallow(<Sidebar />);
        expect(SidebarItem.html()).toMatchSnapshot();
    });

    it('should call useEffect and handleAdministrate func', async () => {
        matchPath.mockImplementation(() => undefined);
        getActiveClientApps.mockResolvedValue(clientAppListTestResponse.list);

        const SidebarItem = render(<Sidebar />);
        expect(getActiveClientApps).toBeCalled();

        await sleep();

        fireEvent.click(SidebarItem.getByText('Витрина ВСП'));
        expect(saveAppCode).toBeCalled();
        expect(goApp).toBeCalled();
    });

    it('should not render Sidebar', () => {
        matchPath.mockImplementation(() => '/admin/users');
        const SidebarItem = shallow(<Sidebar />);
        expect(SidebarItem.html()).toBeNull();
    });

    it('should render menu and return undefined in useEffect', () => {
        getRole.mockImplementation(() => 'UserManager');
        getAppCode.mockImplementation(() => 'mobile-sales-manager');
        menuRoleFuncs.resolveMenuItemsByRoleAndAppCode = jest.fn(
            () => resolveMenuItemsByRoleAndAppCodeValue
        );

        const { container } = render(<Sidebar />);
        expect(container.getElementsByClassName('menu')[0]).toBeInTheDocument();
        expect(getActiveClientApps).toBeCalledTimes(0);
    });

    it('should not render menuItems', () => {
        menuRoleFuncs.resolveMenuItemsByRoleAndAppCode.mockImplementation(() => [ [] ]);
        const SidebarItem = shallow(<Sidebar />);
        expect(SidebarItem.find('Menu.Item')).toHaveLength(0);
    });

    it('should not render appLink', () => {
        matchPath.mockImplementation(() => undefined);
        getActiveClientApps.mockRejectedValue([]);
        menuRoleFuncs.resolveMenuItemsByRoleAndAppCode = jest.fn(
            () => resolveMenuItemsByRoleAndAppCodeValue
        );

        const SidebarItem = shallow(<Sidebar />);
        expect(SidebarItem.find('.appLink')).toHaveLength(0);
    });
});
