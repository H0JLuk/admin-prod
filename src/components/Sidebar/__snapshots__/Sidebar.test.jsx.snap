// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> tests Sidebar snapshot 1`] = `"<div class=\\"menu\\"><div class=\\"blockLogo\\"><img class=\\"logo\\" src=\\"sber-logo.png\\" alt=\\"sberLogo\\"/></div><ul class=\\"ant-menu ant-menu-light ant-menu-root ant-menu-inline\\" role=\\"menu\\"></ul></div>"`;
